source/fe/mapping_q1.cc

```diff
@@ -1705,73 +1703,171 @@ transform_real_to_unit_cell_internal
-  Point<spacedim> p_real = transform_unit_to_real_cell_internal(mdata);
-  Tensor<1,spacedim> f = p_real-p;
-
-  // early out if we already have our point
-  if (f.norm_square() < 1e-24 * cell->diameter() * cell->diameter())
-    return p_unit;
-
-  // we need to compare the position of the computed p(x) against the given
-  // point 'p'. We will terminate the iteration and return 'x' if they are
-  // less than eps apart. The question is how to choose eps -- or, put maybe
-  // more generally: in which norm we want these 'p' and 'p(x)' to be eps
-  // apart.
-  //
-  // the question is difficult since we may have to deal with very elongated
-  // cells where we may achieve 1e-12*h for the distance of these two points
-  // in the 'long' direction, but achieving this tolerance in the 'short'
-  // direction of the cell may not be possible
-  //
-  // what we do instead is then to terminate iterations if
-  //    \| p(x) - p \|_A < eps
-  // where the A-norm is somehow induced by the transformation of the cell.
-  // in particular, we want to measure distances relative to the sizes of
-  // the cell in its principal directions.
-  //
-  // to define what exactly A should be, note that to first order we have
-  // the following (assuming that x* is the solution of the problem, i.e.,
-  // p(x*)=p):
-  //    p(x) - p = p(x) - p(x*)
-  //             = -grad p(x) * (x*-x) + higher order terms
-  // This suggest to measure with a norm that corresponds to
-  //    A = {[grad p(x]^T [grad p(x)]}^{-1}
-  // because then
-  //    \| p(x) - p \|_A  \approx  \| x - x* \|
-  // Consequently, we will try to enforce that
-  //    \| p(x) - p \|_A  =  \| f \|  <=  eps
-  //
-  // Note that using this norm is a bit dangerous since the norm changes
-  // in every iteration (A isn't fixed by depends on xk). However, if the
-  // cell is not too deformed (it may be stretched, but not twisted) then
-  // the mapping is almost linear and A is indeed constant or nearly so.
-  const double eps = 1.e-11;
-  const unsigned int newton_iteration_limit = 20;
-
-  unsigned int newton_iteration = 0;
-  double last_f_weighted_norm;
-  do
+  for(unsigned int newton_iteration = 0; newton_iteration < newton_iteration_limit; newton_iteration++) 
     {
-#ifdef DEBUG_TRANSFORM_REAL_TO_UNIT_CELL
-      std::cout << "Newton iteration " << newton_iteration << std::endl;
-#endif
-
-      // f'(x)
-      Tensor<2,spacedim> df;
+  
+  // shape values and derivatives at x_unit_n
+  compute_shapes(std::vector<Point<dim> > (1, x_unit_n), mdata);
+  
+  // p(x_unit_n)    
+  p_real_n= transform_unit_to_real_cell_internal(mdata);
+      
+      // Initialize the Jacobian
+      Tensor<2, spacedim> J;
+      
+/*     
+  TODO: Would this be more efficient/cheaper?
+  
+   for (unsigned int i=0; i<spacedim; ++i)
+            for (unsigned int j=0; j<dim; ++j)
+              J[i][j]=0;
+	    
+	    or, perhaps
+	    
+    J.clear();
```

Comment:
No, you can't beat the initialization of a local variable on the stack. Internally, there's of course some kind of double loop, but because you limit the lifetime of the variable to the local scope, the compiler knows the values of the tensor elements and can forward-propagate these values. It would also be a premature optimization to worry about the initialization of a local variable unless you have evidence of it being a bottleneck.

So just remove the code.